dataset/src/main/scala/frameless/ops/MatchesAll.scala

```diff
@@ -0,0 +1,19 @@
+package frameless
+package ops
+
+import shapeless.HList
+import shapeless.ops.hlist.Filter
+
+import scala.annotation.implicitNotFound
+
+
+@implicitNotFound("HList of type '${L}' is NOT comprised of only elements of type '${U}'")
+trait MatchesAll[L <: HList, U]
+
+object MatchesAll {
+  implicit def matchesAll[L <: HList, U, Rest <: HList]
+  (implicit
+   filtered: Filter.Aux[L, U, Rest],
+   remainedSame: Rest =:= L
```

Comment:
I am thinking about this one ... So `MatchesAll[A,B]` exists only if A is an `HList` and all the elements of the `HList` are of type `B`. FilterNot doesn't exactly have the same semantics. I can probably implement MatchesAll in terms of FilterNot. 

Essentially MatchesAll breaks down the complexity of Pivot's implicit resolution. Like in the values space you break down complex functions into smaller reusable functions, here I broke down that complex pivot inductive step into MatchesAll, Repeat, etc.